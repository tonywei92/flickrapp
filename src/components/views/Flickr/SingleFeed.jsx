/* eslint-disable camelcase */
import React, { forwardRef } from 'react';
import PropTypes from 'prop-types';

// import material-ui components
import { makeStyles } from '@material-ui/core/styles';
import Card from '@material-ui/core/Card';
import CardActionArea from '@material-ui/core/CardActionArea';
import CardActions from '@material-ui/core/CardActions';
import CardContent from '@material-ui/core/CardContent';
import Link from '@material-ui/core/Link';
import CardMedia from '@material-ui/core/CardMedia';
import Button from '@material-ui/core/Button';
import Typography from '@material-ui/core/Typography';

const { REACT_APP_FLICKR_IMAGE_ERROR_FALLBACK } = process.env;

const useStyles = makeStyles({
  media: {
    // limit image height
    height: 'calc( 100vh - 400px )',
  },
  textEllipsis: {
    textOverflow: 'ellipsis',
    overflow: 'hidden',
    whiteSpace: 'nowrap',
  },
});

const imageErrorFallback = REACT_APP_FLICKR_IMAGE_ERROR_FALLBACK;

const SingleFeedsView = forwardRef((props, ref) => {
  const {
    title,
    author,
    categories,
    direct_link,
    flickr_link,
    published_at,
  } = props.feed;
  const classes = useStyles();

  const openInFlickr = () => {
    window.open(flickr_link, 'blank');
  };

  const setImageFallback = (eventObj) => {
    // eslint-disable-next-line no-param-reassign
    eventObj.target.src = imageErrorFallback;
  };

  return (
    <Card ref={ref} variant="elevation" elevation={8}>
      <CardActionArea onClick={openInFlickr}>
        <CardMedia
          className={classes.media}
          src={direct_link}
          component="img"
          title={title}
          onError={(e) => setImageFallback(e)}
        />
        <CardContent>
          <Typography
            gutterBottom
            variant="h5"
            component="h2"
            className={classes.textEllipsis}
            title={title}
          >
            {title.trim() === '' ? <i>No title</i> : title}
          </Typography>
          <Typography
            variant="body2"
            color="textSecondary"
            component="p"
            className={classes.textEllipsis}
          >
            Posted by&nbsp;
            <Link href={author.uri} target="blank">
              {author.name}
            </Link>
            &nbsp;at
            {published_at}
          </Typography>
          <Typography
            variant="body2"
            color="textSecondary"
            component="p"
            className={classes.textEllipsis}
            title={categories.join(', ')}
          >
            Categories:&nbsp;
            {categories.join('') ? categories.join(', ') : <i>Uncategorized</i>}
          </Typography>
        </CardContent>
      </CardActionArea>

      <CardActions>
        <Button size="small" color="primary">
          Share
        </Button>
        <Button size="small" color="primary" href={flickr_link} target="blank">
          Open in Flickr
        </Button>
      </CardActions>
    </Card>
  );
});

SingleFeedsView.propTypes = {
  feed: PropTypes.shape({
    title: PropTypes.string.isRequired,
    author: PropTypes.shape({
      name: PropTypes.string,
      uri: PropTypes.string,
    }).isRequired,
    categories: PropTypes.arrayOf(PropTypes.string).isRequired,
    direct_link: PropTypes.string.isRequired,
    flickr_link: PropTypes.string.isRequired,
    published_at: PropTypes.string.isRequired,
  }).isRequired,
};

export default SingleFeedsView;
