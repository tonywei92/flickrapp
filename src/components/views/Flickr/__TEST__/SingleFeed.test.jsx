import React from 'react';
import renderer from 'react-test-renderer';
import SingleFeed from '../SingleFeed';

const responseMock = require('./flickrResponse.json');

test('Render correct single feed structure', () => {
  const component = renderer.create(<SingleFeed feed={responseMock[0]} />);
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
