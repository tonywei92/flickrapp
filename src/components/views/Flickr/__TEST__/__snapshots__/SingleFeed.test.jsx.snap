// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render correct single feed structure 1`] = `
<div
  className="MuiPaper-root MuiCard-root MuiPaper-elevation8 MuiPaper-rounded"
>
  <button
    className="MuiButtonBase-root MuiCardActionArea-root"
    disabled={false}
    onBlur={[Function]}
    onClick={[Function]}
    onDragLeave={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    tabIndex={0}
    type="button"
  >
    <img
      className="MuiCardMedia-root makeStyles-media-1 MuiCardMedia-media MuiCardMedia-img"
      onError={[Function]}
      src="https://live.staticflickr.com/65535/50205609033_bffac90b1c_b.jpg"
      title="Golden hour"
    />
    <div
      className="MuiCardContent-root"
    >
      <h2
        className="MuiTypography-root makeStyles-textEllipsis-2 MuiTypography-h5 MuiTypography-gutterBottom"
        title="Golden hour"
      >
        Golden hour
      </h2>
      <p
        className="MuiTypography-root makeStyles-textEllipsis-2 MuiTypography-body2 MuiTypography-colorTextSecondary"
      >
        Posted by 
        <a
          className="MuiTypography-root MuiLink-root MuiLink-underlineHover MuiTypography-colorPrimary"
          href="https://www.flickr.com/people/gio_vincenti/"
          onBlur={[Function]}
          onFocus={[Function]}
          target="blank"
        >
          Giovanni Vincenti
        </a>
         at
        2020-08-09T13:48:19Z
      </p>
      <p
        className="MuiTypography-root makeStyles-textEllipsis-2 MuiTypography-body2 MuiTypography-colorTextSecondary"
        title=""
      >
        Categories: 
        <i>
          Uncategorized
        </i>
      </p>
    </div>
    <span
      className="MuiCardActionArea-focusHighlight"
    />
  </button>
  <div
    className="MuiCardActions-root MuiCardActions-spacing"
  >
    <button
      className="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-textSizeSmall MuiButton-sizeSmall"
      disabled={false}
      onBlur={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={0}
      type="button"
    >
      <span
        className="MuiButton-label"
      >
        Share
      </span>
    </button>
    <a
      aria-disabled={false}
      className="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-textSizeSmall MuiButton-sizeSmall"
      href="https://www.flickr.com/photos/gio_vincenti/50205609033/"
      onBlur={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={0}
      target="blank"
    >
      <span
        className="MuiButton-label"
      >
        Open in Flickr
      </span>
    </a>
  </div>
</div>
`;
